connexion/middleware/abstract.py

```diff
@@ -1,42 +1,186 @@
 import abc
 import logging
 import pathlib
+import sys
 import typing as t
 
-import typing_extensions as te
 from starlette.types import ASGIApp, Receive, Scope, Send
 
-from connexion.apis.abstract import AbstractSpecAPI
-from connexion.exceptions import MissingMiddleware
+from connexion.exceptions import MissingMiddleware, ResolverError
 from connexion.http_facts import METHODS
+from connexion.jsonifier import Jsonifier
 from connexion.operations import AbstractOperation
-from connexion.resolver import ResolverError
+from connexion.resolver import Resolver
+from connexion.spec import Specification
 
-logger = logging.getLogger("connexion.middleware.abstract")
+logger = logging.getLogger(__name__)
 
 ROUTING_CONTEXT = "connexion_routing"
 
 
-class AppMiddleware(abc.ABC):
-    """Middlewares that need the APIs to be registered on them should inherit from this base
-    class"""
+class SpecMiddleware(abc.ABC):
+    """Middlewares that need the specification(s) to be registered on them should inherit from this
+    base class"""
 
     @abc.abstractmethod
     def add_api(
         self, specification: t.Union[pathlib.Path, str, dict], **kwargs
     ) -> None:
+        """
+        Register an API represented by a single OpenAPI specification on this middleware.
+        Multiple APIs can be registered on a single middleware.
+        """
+
+    @abc.abstractmethod
+    async def __call__(self, scope: Scope, receive: Receive, send: Send):
         pass
 
 
-class RoutedOperation(te.Protocol):
-    def __init__(self, next_app: ASGIApp, **kwargs) -> None:
-        ...
+class AbstractSpecAPI:
+    """Base API class with only minimal behavior related to the specification."""
 
-    async def __call__(self, scope: Scope, receive: Receive, send: Send) -> None:
-        ...
+    jsonifier = Jsonifier()
+
+    def __init__(
+        self,
+        specification: t.Union[pathlib.Path, str, dict],
+        base_path: t.Optional[str] = None,
+        resolver: t.Optional[Resolver] = None,
+        arguments: t.Optional[dict] = None,
+        uri_parser_class=None,
+        *args,
+        **kwargs,
+    ):
+        self.specification = Specification.load(specification, arguments=arguments)
+        self.uri_parser_class = uri_parser_class
+
+        self._set_base_path(base_path)
+
+        self.resolver = resolver or Resolver()
+
+    def _set_base_path(self, base_path: t.Optional[str] = None) -> None:
+        if base_path is not None:
+            # update spec to include user-provided base_path
+            self.specification.base_path = base_path
+            self.base_path = base_path
+        else:
+            self.base_path = self.specification.base_path
+
+
+OP = t.TypeVar("OP")
+""""""
```

Comment:
A random string appeared? 😁

Reply:
Woops, seems like I forgot to push my latest fixes.